### PR TITLE
Adding a minimal implementation of std::optional.

### PR DESCRIPTION
**Pull request by spielawa**
We're using C++11, which doesn't have std::optional. We need a few
features of std::optional for improved GPS support. These are
implemented here.

### CODE REVIEW

File: cartographer/common/optional.h [hunk 1]
```diff
@@ -0,0 +1,39 @@
+#ifndef CARTOGRAPHER_COMMON_OPTIONAL_H_
```
Comment (wohe) [inline cartographer/common/optional.h]: This needs a license comment. Same with the other file.

Comment (spielawa) [reply cartographer/common/optional.h]: done

**Comment by spielawa** [comment]
I signed it!

**Comment by SirVer** [comment]
@googlebot ping?

**Comment by cschuet** [comment]
@SirVer @spielawa We checked the Google CLA database and Github user spielawa has not signed the CLA yet. Can you double check that you actually specified the right email address when signing?

**Comment by spielawa** [comment]
I signed it!

**Comment by wohe** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user wohe. Merge queue now has a length of 1.

**Pull request merged**